lib/azure/armrest/armrest_manager.rb

```diff
@@ -209,14 +209,71 @@ def providers
 
       # Returns information about the specific provider +namespace+.
       #
-      def provider_info(namespace)
-        url = url_with_api_version(@base_url, 'providers', namespace)
+      def provider_info(provider)
+        url = url_with_api_version(@base_url, 'providers', provider)
         response = rest_get(url)
         JSON.parse(response.body)
       end
 
       alias geo_locations provider_info
 
+      # Returns a list of all locations for all resource types of the given
+      # +provider+. If you do not specify a provider, then the locations for
+      # all providers will be returned.
+      #
+      # If you need individual details on a per-provider basis, use the
+      # provider_info method instead.
+      #--
+      # TODO: I think a 7-day cache may be wise here as the overall list of
+      # locations is unlikely to change from day to day.
+      #
+      def locations(provider = nil)
+        array = []
+
+        if provider
+          info = provider_info(provider)
+          info['resourceTypes'].map{ |rt| array << rt['locations'] }
```

Comment:
So, the main difference is that `collect` (or `map`) is already instantiating a new array and appending to it implicitly.  So, by appending the value to your own array in the block, you're bypassing the advantage of using `collect`.